synapse/storage/state.py

```diff
@@ -48,6 +50,179 @@ def __len__(self):
         return len(self.delta_ids) if self.delta_ids else 0
 
 
+@attr.s(slots=True)
+class StateFilter(object):
+    """A filter used when querying for state.
+
+    The default, `StateFilter()`, is equivalent to no filter.
+
+    Attributes:
+        types (Iterable[str, str|None]|None): list of 2-tuples of the form
```

Comment:
afaict we iterate over this stuff multiple times, so a plain iterable won't do. s/Iterable/list/, possibly.